idea/src/org/jetbrains/kotlin/idea/editor/KotlinLiteralCopyPasteProcessor.kt

```diff
@@ -0,0 +1,285 @@
+
+        for (i in startOffsets.indices) {
+            if (i > 0) {
+                buffer.append('\n') // LF is added for block selection
+            }
+
+            val fileRange = TextRange(startOffsets[i], endOffsets[i])
+
+            var givenTextOffset = fileRange.startOffset
+            while (givenTextOffset < fileRange.endOffset) {
+                val element: PsiElement? = file.findElementAt(givenTextOffset)
+                if (element == null) {
+                    buffer.append(fileText.substring(givenTextOffset, fileRange.endOffset))
+                    break
+                }
+                val elTp = element.node.elementType
+                if (elTp == KtTokens.ESCAPE_SEQUENCE && fileRange.contains(element.range) && !fileRange.contains(element.findContainingTemplate().range)) {
+                    val tpEntry = element.parent as KtEscapeStringTemplateEntry
+                    changed = true
+                    buffer.append(tpEntry.unescapedValue)
+                    givenTextOffset = element.endOffset
+                }
+                else if (elTp == KtTokens.SHORT_TEMPLATE_ENTRY_START || elTp == KtTokens.LONG_TEMPLATE_ENTRY_START) {
+                    //Process inner templates without escaping
+                    val tpEntry = element.parent
+                    val inter = fileRange.intersection(tpEntry.range)!!
+                    buffer.append(fileText.substring(inter.startOffset, inter.endOffset))
+                    givenTextOffset = inter.endOffset
+                }
+                else {
+                    val inter = fileRange.intersection(element.range)!!
+                    buffer.append(fileText.substring(inter.startOffset, inter.endOffset))
+                    givenTextOffset = inter.endOffset
+                }
+
+            }
+            val blockSelectionPadding = deducedBlockSelectionWidth - fileRange.length
+            for (j in 0..blockSelectionPadding - 1) {
+                buffer.append(' ')
+            }
+
+        }
+
+        return if (changed) buffer.toString() else null
+    }
+
+    override fun preprocessOnPaste(project: Project, file: PsiFile, editor: Editor, text: String, rawText: RawText?): String {
+        if (file !is KtFile){
+            return text
+        }
+        PsiDocumentManager.getInstance(project).commitDocument(editor.document)
+        val selectionModel = editor.selectionModel
+        val beginTp = file.getTemplateIfAtLiteral(selectionModel.selectionStart) ?: return text
+        val endTp = file.getTemplateIfAtLiteral(selectionModel.selectionEnd) ?: return text
+        if (beginTp.isSingleQuoted() != endTp.isSingleQuoted()) {
+            return text
+        }
+
+
+        return if (beginTp.isSingleQuoted()) {
+            TemplateTokenizer(text).map {
+                when (it) {
+                    is LiteralChunk -> StringUtil.escaper(true, "\$\"").`fun`(it.text)
```

Comment:
I think that it's better to create a single `StringBuilder` to hold the entire preprocessed contents, and to use the overload of `escapeStringCharacters` that takes a `StringBuilder` as a parameter.